src/DynamicData/Cache/Internal/TransformOnObservable.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) 2011-2023 Roland Pheasant. All rights reserved.
+// Roland Pheasant licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for full license information.
+
+using System.Reactive.Disposables;
+using System.Reactive.Linq;
+using DynamicData.Internal;
+using DynamicData.Kernel;
+
+namespace DynamicData.Cache.Internal;
+
+internal sealed class TransformOnObservable<TSource, TKey, TDestination>(IObservable<IChangeSet<TSource, TKey>> source, Func<TSource, TKey, IObservable<TDestination>> transform)
+    where TSource : notnull
+    where TKey : notnull
+    where TDestination : notnull
+{
+    public IObservable<IChangeSet<TDestination, TKey>> Run() => Observable.Create<IChangeSet<TDestination, TKey>>(observer =>
+    {
+        var locker = new object();
+        var cache = new ChangeAwareCache<TDestination, TKey>();
+        var updateCounter = 0;
+
+        // Helper to emit any pending changes when all the updates have been handled
+        void EmitChanges()
+        {
+            if (Interlocked.Decrement(ref updateCounter) == 0)
```

Comment:
I think the concept of an update counter to batch results is inspired.  The same concept would probably benefit several other operations.

Reply:
> I think the concept of an update counter to batch results is inspired. The same concept would probably benefit several other operations.

Thank you!  :blush:
I used it for `TransformManyAsync` to get behavior similar to `TransformMany` after you stressed the importance of minimizing the number of ChangeSets.  I was also thinking that it might be beneficial to use in other places, so I'll keep a lookout for those.

There is some nuance to using it.  For example, you have to increment the counter before the `Synchronize` so that updates coming in on other threads are counted and the emission downstream is deferred until it can be handled.

There are also some downsides:
- More difficult (sometimes impossible) to deterministically know how many changesets you'll get
- Possible to never get a downstream changeset (if updates happen constantly)

Reply:
Thank you!  :blush:
I used it for `TransformManyAsync` to get behavior similar to `TransformMany` whenever possible.  I was also thinking that it might be beneficial to other places.  Perhaps it warrants an abstraction that makes it easier to integrate...  But then again, perhaps not.